### PR TITLE
dlink: dial nodes concurrently

### LINKED ISSUE #3: [dlink] dial nodes concurrently
**Describe the feature or problem you'd like to solve**
Currently the available nodes are dialed one by one when establishing connection to the nodes. This stalls the process till all the nodes are dialled:
https://github.com/pee2pee/pee2pee/blob/main/dlink/piecestore.go#L50-L58

**Propose a Solution**
Dial the available nodes concurrently like we do when sending requests to the nodes (for uploads, downloads, etc.)

**Additional context**

### PR DESCRIPTION
**Pull request by Bostigger**
## Description
This commit fixes concurrency in dialing the nodes individually

## Related Issue

Resolves #3

## Types of changes

- [✔ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: dlink/piecestore.go [hunk 1]
```diff
@@ -47,18 +49,31 @@ func NewPieceStore(ctx context.Context, config *Config, log *zap.Logger) (*Piece
...
+		wg.Add(1)
+		go func(node DataNode) {
+			defer wg.Done()
+			conn, err := dialNode(ctx, node.Address, log)
+			if err != nil {
+				log.Warn("failed to dial node", zap.String("node", node.Address), zap.String("NodeID", node.PeerID.String()), zap.Error(err))
+				return
+			}
+			client := nodeAPIV1.NewDRPCPiecestoreClient(drpcconn.New(conn))
+			p.preventRace(client)
```
Comment (profclems) [inline dlink/piecestore.go:64]: Instead of creating a new variable, we can just do:
```go
p.preventRace(nodeAPIV1.NewDRPCPiecestoreClient(drpcconn.New(conn)))
```

File: dlink/piecestore.go [hunk 2]
```diff
@@ -47,18 +49,31 @@ func NewPieceStore(ctx context.Context, config *Config, log *zap.Logger) (*Piece
...
+		}(node)
 	}
+	wg.Wait()
 	return p, nil
 }
 
+func (p *PieceStore) preventRace(client nodeAPIV1.DRPCPiecestoreClient) {
```
Comment (profclems) [inline dlink/piecestore.go:71]: `preventRace` is quite a broad name. Should we rename to `addDRPCClient` to make it more specific to what it does?

**Review by profclems — CHANGES_REQUESTED** [review]

File: dlink/piecestore.go (hunk 1 above)
Comment (Bostigger) [reply dlink/piecestore.go:64]: Alright well noted

File: dlink/piecestore.go (hunk 2 above)
Comment (Bostigger) [reply dlink/piecestore.go:71]: That would be great

File: dlink/piecestore.go [hunk 3]
```diff
@@ -47,18 +49,31 @@ func NewPieceStore(ctx context.Context, config *Config, log *zap.Logger) (*Piece
...
+		wg.Add(1)
+		go func(node DataNode) {
+			defer wg.Done()
+			conn, err := dialNode(ctx, node.Address, log)
+			if err != nil {
+				log.Warn("failed to dial node", zap.String("node", node.Address), zap.String("NodeID", node.PeerID.String()), zap.Error(err))
+				return
+			}
+			p.addDRPCClient(nodeAPIV1.NewDRPCPiecestoreClient(drpcconn.New(conn)))
+
```
Comment (profclems) [inline dlink/piecestore.go:64]: nit: unnecessary whitespace

**Review by profclems — APPROVED** [review]
LGTM, just one nit.

**Review by bhrayhaw — APPROVED** [review]

File: dlink/piecestore.go (hunk 3 above)
Comment (Bostigger) [reply dlink/piecestore.go:64]: Ok sure!
